shell/platform/windows/window_state.h

```diff
@@ -36,8 +38,31 @@ struct FlutterDesktopPluginRegistrar {
 // Wrapper to distinguish the messenger ref from the engine ref given out
 // in the C API.
 struct FlutterDesktopMessenger {
+ public:
+  FlutterDesktopMessenger() = default;
+  flutter::FlutterWindowsEngine* GetEngine() const { return engine; }
+  void SetEngine(flutter::FlutterWindowsEngine* arg_engine) {
+    std::scoped_lock lock(mutex_);
+    engine = arg_engine;
+  }
+  FlutterDesktopMessenger* AddRef() {
+    ref_count_.fetch_add(1);
+    return this;
+  }
+  void Release() {
+    int32_t old_count = ref_count_.fetch_sub(1);
+    if (old_count <= 1) {
+      delete this;
```

Comment:
The lifetimes of this object as its used seems incredibly confusing now, unless I'm missing something. It's being kept in a ref-counted pointer, which one would expect to maintain the lifetime tracking, but it then has this other internal reference system that can cause it to self-delete and leave *dangling* ref-counting pointers? Am I misunderstanding?

Reply:
Yea, I think you are missing the code where the shared_ptr is constructed.  It is retaining and releasing the object's internal ref count, not using the one in the smart pointer.  This is why I mentioned elsewhere that we could still use a unique_ptr if we wanted but it is confusing since it isn't unique.  You can't get a dangling shared_ptr.